### ISSUE TITLE
Inconsistency with setItemIndex and Navigator / World state

### ISSUE THREAD

**Issue by thec0keman** [openseadragon/openseadragon#1843]
When calling `setItemIndex` inside a `add-item` callback, there is an inconsistency in the World and Navigator states.  The former has the item that triggered the `add-item` callback, but World does not.

So, given this use case:
- add an item with viewer.addTiledImage
- code in `add-item` callback inspects world._items and determines that one of the items needs to be moved to top of the list, so it calls `setItemIndex`
- `setItemIndex` updates World state, and dispatches the `item-index-change` event
- `item-index-change` fires.  

However, since the Navigator state does not yet have the item from `add-item`, an error is thrown:
```
Uncaught Error: Index bigger than number of layers.
```

I'm not sure if this matters in our use case, since the Navigator state eventually does include the item, it is just in a different ordering than World.  However, I think the expected behavior would be that state for both would be updated when the `add-item` callback fires.

**Comment by RammasEchor** [comment]
Hi! I'm interested in this issue, may I give it a try?

**Comment by iangilman** [comment]
@RammasEchor Absolutely, please do! Let me know if you have any questions along the way :-)

**Comment by RammasEchor** [comment]
Thank you! I do have a question; I've been reading the code, and I have been trying to replicate the error with this code:

```js
viewer.world.addHandler('add-item', e => {
    viewer.world.setItemIndex(e.item, 0);
    console.log(viewer.world.getItemCount(e.item));
});

viewer.addTiledImage({
    tileSource: source,
});
```
<p>
However, with this code the error 
</p>

```
Uncaught Error: Index bigger than number of layers.
```

is non deterministic; I can't hit it regularly. With this code I get another type of error, which is two assertions about an setting an index of an undefined item.
The expected behavior of the above code is to set any item that is added to the `0` index (so, behind all other items), however, the view in the navigator is not updated, since the world belonging to the nevigator does not have the items updated at the moment of `setItemIndex`.
My questions are:

- Do I create another issue about the assertions? This seems closely related to this one, so I don't think it would be necessary. I do have a solution for those.
- Is there a way to consistently reproduce the original error? I tried various thing, but I do think I'm missing something important.

Thank you very much!

**Comment by iangilman** [comment]
Now that I'm taking a closer look at the code, I think the root of the issue may be that we are adding the new image to the world: 

https://github.com/openseadragon/openseadragon/blob/577622a210956b0ed84b25f2e5f15b619aedc64a/src/viewer.js#L1505

...before we add it to the navigator:

https://github.com/openseadragon/openseadragon/blob/577622a210956b0ed84b25f2e5f15b619aedc64a/src/viewer.js#L1525

It might be worth flipping that order to see if that improves things (and makes them more reliable). 

As for fixing the other issue you found, it seems reasonable to do it all at the same time, since it's in the same area. I'd love to see the code change; I'm having a hard time visualizing what you've done. 

Thank you for digging into this!

**Comment by RammasEchor** [comment]
Yes of course! I will create a draft pull request with my work. I would greatly appreciate any feedback.

**Comment by iangilman** [comment]
@RammasEchor excellent... looks good to me!

**Comment by iangilman** [comment]
Fixed by #2005
